cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java

```diff
@@ -336,9 +338,112 @@ public void rollback(TopicId topicId, RollbackDetail rollbackDetail)
+   *             - Upon reading rows, chunked messages are reconstructed as a single message and are held in memory.
+   *         </li>
+   *     </ul>
+   * </p>
+   */
   @Override
   public CloseableIterator<RawMessage> fetch(MessageFetchRequest messageFetchRequest)
       throws TopicNotFoundException, IOException {
-    throw new IOException("NOT IMPLEMENTED");
+    // If the client is fetching for the first time ever,
+    // they should fetch data for publish_ts > 0 and sequence_id > -1.
+    Long startTime = 0L;
+    if (messageFetchRequest.getStartTime() != null) {
+      startTime = messageFetchRequest.getStartTime();
+    }
+    short sequenceId = -1;
+    // MessageId is constructed as the concatenation of SequenceNo and CommitTs columns.
+    // These fields are set in the offset field of the message fetch request.
+    byte[] id = messageFetchRequest.getStartOffset();
+    if (id != null) {
+      int offset = 0;
+      startTime = Bytes.toLong(id, offset);
+      offset += Bytes.SIZEOF_LONG;
+      sequenceId = Bytes.toShort(id, offset);
+    }
+
+    LOG.trace("Fetching message from topic : {} with start time : {} sequenceId : {}",
+        messageFetchRequest.getTopicId().getTopic(), startTime, sequenceId);
+    // publish_ts > TIMESTAMP_MICROS(startTime)
+    // or
+    // publish_ts = TIMESTAMP_MICROS(startTime) and sequence_id > sequenceId
+    // order by
+    // publish_ts, sequence_id
+    String sqlStatement = String.format(
+        "SELECT %s, %s, UNIX_MICROS(%s), %s FROM %s where (%s > TIMESTAMP_MICROS(%s)) or"
+            + " (%s = TIMESTAMP_MICROS(%s) and %s > %s) order by" + " %s, %s LIMIT %s",
+        SpannerMessagingService.SEQUENCE_ID_FIELD,
+        SpannerMessagingService.PAYLOAD_SEQUENCE_ID_FIELD,
+        SpannerMessagingService.PUBLISH_TS_FIELD, SpannerMessagingService.PAYLOAD_FIELD,
+        SpannerMessagingService.getTableName(messageFetchRequest.getTopicId()),
+        PUBLISH_TS_FIELD, startTime, PUBLISH_TS_FIELD, startTime,
+        SEQUENCE_ID_FIELD, sequenceId, PUBLISH_TS_FIELD, SEQUENCE_ID_FIELD,
+        messageFetchRequest.getLimit());
+
+    try {
+      ResultSet resultSet = client.singleUse().executeQuery(Statement.of(sqlStatement));
+      return new SpannerResultSetClosableIterator<>(resultSet);
+    } catch (Exception ex) {
+      throw new IOException(ex);
+    }
+  }
+
+  public static class SpannerResultSetClosableIterator<RawMessage> extends
+      AbstractCloseableIterator<io.cdap.cdap.messaging.spi.RawMessage> {
+
+    private final ResultSet resultSet;
+
+    public SpannerResultSetClosableIterator(ResultSet resultSet) {
+      this.resultSet = resultSet;
+    }
+
+    @Override
+    protected io.cdap.cdap.messaging.spi.RawMessage computeNext() {
+      if (!resultSet.next()) {
+        return endOfData();
+      }
+
+      byte[] id = getMessageId(resultSet.getLong(0), resultSet.getLong(1), resultSet.getLong(2));
```

Comment:
Please use constants for field indexes. Or if spanner supports retrieval by fields names, use that